### PR TITLE
Allow /balancetop to hide players with a balance of 0

### PR DESCRIPTION
**Pull request by DebugOk**
### Information

This PR closes [discussion 4205](https://github.com/EssentialsX/Essentials/discussions/4205). 

### Details

**Proposed feature:**  

Add the ability to hide players that have a balance of 0 from /balancetop

**Environments tested:**    

OS: Linux 4.15.0-106-generic (amd64)

Java version:  OpenJDK 64-Bit Server VM 25.212-b03

- [x] Most recent Paper version (1.16.5, git-Paper-776)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

### CODE REVIEW

**Comment by DebugOk** [comment]
I'm looking into optimizing, however I'm not sure how I would be able to do that

**Comment by DebugOk** [comment]
Previews

Preview of changes to config.yml
```
# Show users with a balance of zero in /balacetop.
# NOTE: This may not apply right away, run '/balancetop force' if it doesn't.
show-zero-topbal: true
```

Videos:
[show-zero-topbal: false (With money)](https://user-images.githubusercontent.com/49997488/121761836-8da2eb80-cb32-11eb-9f88-6ad1b417bb5f.mp4)
[show-zero-topbal: false (Without money)](https://user-images.githubusercontent.com/49997488/121761838-8ed41880-cb32-11eb-8731-1ab172880d4f.mp4)
[show-zero-topbal: true (With money)](https://user-images.githubusercontent.com/49997488/121761839-909ddc00-cb32-11eb-9618-2eee911c1a2c.mp4)
[show-zero-topbal: true (Without money)](https://user-images.githubusercontent.com/49997488/121761840-9398cc80-cb32-11eb-9f34-cbf532d42e12.mp4)

File: Essentials/src/main/resources/config.yml [hunk 1]
```diff
@@ -781,6 +781,10 @@ minimum-pay-amount: 0.001
 # Enable this to block users who try to /pay another user which ignore them.
 pay-excludes-ignore-list: false
 
+# Show users with a balance of zero in /balacetop.
+# NOTE: This may not apply right away, run '/balancetop force' if it doesn't.
+show-zero-topbal: true
```
Comment (JRoy) [inline Essentials/src/main/resources/config.yml]:
```suggestion
# Whether or not users with a balance less than or equal to $0 should be shown in balance-top.
# Setting to false will not show people with balances <= 0 in balance-top.
# NOTE: After reloading the config, you must also run '/baltop force' for this to appear
show-zero-baltop: true
```

File: Essentials/src/main/java/com/earth2me/essentials/Settings.java [hunk 2]
```diff
@@ -1858,4 +1858,9 @@ public boolean isRespawnAtBed() {
     public boolean isUpdateCheckEnabled() {
         return config.getBoolean("update-check", true);
     }
+
+    @Override
+    public boolean showZeroTopBal() {
+        return config.getBoolean("show-zero-topbal", true);
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/Settings.java]:
```suggestion
        return config.getBoolean("show-zero-baltop", true);
```

File: Essentials/src/main/java/com/earth2me/essentials/Settings.java [hunk 3]
```diff
@@ -1858,4 +1858,9 @@ public boolean isRespawnAtBed() {
     public boolean isUpdateCheckEnabled() {
         return config.getBoolean("update-check", true);
     }
+
+    @Override
+    public boolean showZeroTopBal() {
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/Settings.java]:
```suggestion
    public boolean showZeroBaltop() {
```

File: Essentials/src/main/java/com/earth2me/essentials/ISettings.java [hunk 4]
```diff
@@ -379,6 +379,8 @@
 
     boolean isUpdateCheckEnabled();
 
+    boolean showZeroTopBal();
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/ISettings.java]:
```suggestion
    boolean showZeroBaltop();
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java [hunk 5]
```diff
@@ -103,7 +103,9 @@ public void run() {
                     newCache.getLines().add(tl("serverTotal", NumberUtil.displayCurrency(ess.getBalanceTop().getBalanceTopTotal(), ess)));
                     int pos = 1;
                     for (final Map.Entry<UUID, BalanceTop.Entry> entry : ess.getBalanceTop().getBalanceTopCache().entrySet()) {
-                        newCache.getLines().add(tl("balanceTopLine", pos, entry.getValue().getDisplayName(), NumberUtil.displayCurrency(entry.getValue().getBalance(), ess)));
+                        if (ess.getSettings().showZeroTopBal() || entry.getValue().getBalance().intValue() > 0) {
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandbalancetop.java]:
```suggestion
                        if (ess.getSettings().showZeroBaltop() || entry.getValue().getBalance().compareTo(BigDecimal.ZERO) <= 0) {
```

**Review by JRoy — APPROVED** [review]

**Comment by DebugOk** [comment]
Looks good. Sorry, I'm not the best at writing good config descriptions ^^

**Comment by JRoy** [comment]
Thanks for the contribution!

**Comment by molor** [comment]
Why we can't filter and remove players in #calculateBalanceTopMapAsync() BEFORE sorting the list and removing players from it later? If I misunderstood how it works, then I'm sorry.

**Comment by DebugOk** [comment]
That would be a good idea, but I haven't really worked with Java or Essential's before. I'll look into that because I'm pretty sure that'd be a better method

**Comment by DebugOk** [comment]
Looking into it, I'm not sure how much this'll improve it, as it still has to check the player before deciding to add them to the list or not. Instead of checking in `commandBalanceTop.java` it'll check in `BalanceTopImpl.java', and I don't really see how moving the check one file will make it run any faster

**Comment by molor** [comment]
I think that if we just don't add a player to the list before sorting it, if its balance <= 0, then this is what will make the subsequent sorting faster. But I haven't looked at how the list of players is built (and where they even come from), so maybe it can't be done otherwise

**Comment by DebugOk** [comment]
This is the code that manages creating the list for /balancetop

If I'm understanding it right the list is generated every time `/balancetop force` is ran or the cache expires. What is generated in `/balancetop` is also cached so it doesn't loop over all players often
```
private void calculateBalanceTopMap() {
        final List<Entry> entries = new LinkedList<>();
        BigDecimal newTotal = BigDecimal.ZERO;
        for (UUID u : ess.getUserMap().getAllUniqueUsers()) {
            final User user = ess.getUserMap().getUser(u);
            if (user != null) {
                if (!ess.getSettings().isNpcsInBalanceRanking() && user.isNPC()) {
                    // Don't list NPCs in output
                    continue;
                }
                if (!user.isBaltopExempt()) {
                    final BigDecimal userMoney = user.getMoney();
                    user.updateMoneyCache(userMoney);
                    newTotal = newTotal.add(userMoney);
                    final String name = user.isHidden() ? user.getName() : user.getDisplayName();
                    entries.add(new BalanceTop.Entry(user.getBase().getUniqueId(), name, userMoney));
                }
            }
        }
        final LinkedHashMap<UUID, Entry> sortedMap = new LinkedHashMap<>();
        entries.sort((entry1, entry2) -> entry2.getBalance().compareTo(entry1.getBalance()));
        for (Entry entry : entries) {
            sortedMap.put(entry.getUuid(), entry);
        }
        topCache = sortedMap;
        balanceTopTotal = newTotal;
        cacheAge = System.currentTimeMillis();
        cacheLock.complete(null);
        cacheLock = null;
    }
```

**Comment by DebugOk** [comment]
From what I can see, the only benefit of moving the check into #calculateBalanceTopMap() is having it run asynchronously

However, I am unable to test if this actually increases performance due to not being able to test it with more players

**Comment by molor** [comment]
Something like this?

```
private void calculateBalanceTopMap() {
	final List<Entry> entries = new LinkedList<>();
	BigDecimal newTotal = BigDecimal.ZERO;
	final boolean showZero = ess.getSettings().showZeroBaltop();

	for (UUID u : ess.getUserMap().getAllUniqueUsers()) {
		final User user = ess.getUserMap().getUser(u);
		if (user != null) {
			if (!ess.getSettings().isNpcsInBalanceRanking() && user.isNPC()) {
				// Don't list NPCs in output
				continue;
			}
			if (user.isBaltopExempt()) {
				continue;
			}

			final BigDecimal userMoney = user.getMoney();

			if (!showZero && userMoney.compareTo(BigDecimal.ZERO) <= 0) {
				continue;
			}

			user.updateMoneyCache(userMoney);
			newTotal = newTotal.add(userMoney);
			final String name = user.isHidden() ? user.getName() : user.getDisplayName();
			entries.add(new BalanceTop.Entry(user.getBase().getUniqueId(), name, userMoney));
		}
	}

	final LinkedHashMap<UUID, Entry> sortedMap = new LinkedHashMap<>();
	entries.sort((entry1, entry2) -> entry2.getBalance().compareTo(entry1.getBalance()));
	for (Entry entry : entries) {
		sortedMap.put(entry.getUuid(), entry);
	}
	topCache = sortedMap;
	balanceTopTotal = newTotal;
	cacheAge = System.currentTimeMillis();
	cacheLock.complete(null);
	cacheLock = null;
}
```

(I don't have the IDE currently, so sorry for possible mistakes)

**Comment by DebugOk** [comment]
I actually made a test branch on my fork that includes what you just wrote, but with less line changes. It works but I don't really notice a diffrent. I'm not able to test it with multiple players, though

https://github.com/DebugOk/Essentials/commit/987c6383827385c6a7253db20d1dc58134cb1b5b

**Comment by DebugOk** [comment]
I can always just open an issue and see what the maintainers think of it

**Comment by molor** [comment]
You can create a build that runs with and without this change and logs time difference to console, then I run it on my another test server with ~5k playerdata files and post a results

**Comment by DebugOk** [comment]
I can try, but it'd take a bit to get working. I'll just add an `Async` argument that decided where the check is ran

**Comment by DebugOk** [comment]
I've got to do some work outside of coding first, I'll have what you requested ready within a day though. Feel free to add me on Discord @ DebugOk#6605

**Comment by DebugOk** [comment]
@molor I've created what you requested. It's not the best code and you shouldn't use it for anything besides this test, but it gets the job done.

I've made it so /balancetop always runs with the force argument, but I added an async one.

Without async it runs like the current dev build of EssentialsX, but output how long it took to the console.
With async it runs with the previously mentioned code in calculateBalanceTopMap()

Please run both `/balancetop` and `/balancetop async` twice and share the results of how long both took
https://github.com/DebugOk/Essentials/tree/2.x-AsyncshowZeroBalTop

**Comment by DebugOk** [comment]
I'm not really expecting there to be much of a diffrence, but it's still good to test it anyways just in case

**Comment by JRoy** [comment]
Sorting in the baltop command itself is perfectly fine, the baltop command keeps its own cache. The reason I didn't want this in the Baltop api cache itself is because in case other plugins don't want to sort away $0 balances

**Comment by DebugOk** [comment]
Like I already said, I don't really see a benefit in moving the sorting. I doubt there will be a noticable change in the time it takes

I've deleted it, just because I don't see a point in keeping it around. This change will break more then it'll do good

**Comment by molor** [comment]
@JRoy & @DebugOk, just so we know what we're giving up on ):

```
[12:57:44] initial userdata (current implementation): 5372
[12:57:46] entries.size() after for() loop: 5372
[12:57:46] took: 1654 ms

[12:58:16] initial userdata (with remove before sorting): 5372
[12:58:16] entries.size() after for() loop: 36
[12:58:16] took: 13 ms
```

**Comment by DebugOk** [comment]
Be sure to run both methods twice. I've found the first time always takes longer with both methods

**Comment by molor** [comment]
Already, I have SSD (that caches files with frequent access) and also faced with this, average difference/error rate in measurements is 25-50 ms

**Comment by DebugOk** [comment]
Still, like Jroy said, filtering in the actual method isn't a good idea because some other things may want it with players with a balance of zero or less. The only way it may not cause issues if there's is a "filterNoBalance" like parameter

**Comment by molor** [comment]
We can add a new method that remove players before sorting (e.g. `calculateBalanceTopMapAsync(boolean withNoBalanced)`) and use it inside EssentialsX by default, but leave old method (without arguments) for other plugins API. Why not?

`+` Command will work faster because no useless operations (sorting) will be performed
`+` Other plugins will not be affected because API isn't changed (just a one new method added)
`-` ??

**Comment by molor** [comment]
I would even make a pull request myself with this change, but I don't know how.

**Comment by DebugOk** [comment]
If you have the repo public I can open a PR for you, but it should just be a button on the PR page

**Comment by molor** [comment]
> If you have the repo public I can open a PR for you, but it should just be a button on the PR page

I don't ever think how to work with git. I know only how to clone repository and compile, bruh

**Comment by DebugOk** [comment]
Do you have the edited code in a private repository? Like in a fork or do you just have it stored locally?

**Comment by molor** [comment]
> Like in a fork or do you just have it stored locally?

Local only.

**Comment by DebugOk** [comment]
If you upload the files you changes as [gists](https://gist.github.com/) I can put them into a repo and open a PR, otherwise you'll have to make one by either forking Essentials or manually making it
image

**Comment by DebugOk** [comment]
Anyone is able to fork a repo and create pull requests btw, there really aren't any requirements to doing either besides knowing how to

**Comment by molor** [comment]
Something like this: https://gist.github.com/molor/177add22396fbbda5c259748058af12f

**Comment by DebugOk** [comment]
I've created a branch on my fork that you can use to open a PR.

You can [go here](https://github.com/EssentialsX/Essentials/compare/2.x...DebugOk:2.x-ExtendedBaltopAPI?expand=1) to open said PR, and you can use this [template](https://raw.githubusercontent.com/EssentialsX/Essentials/2.x/.github/PULL_REQUEST_TEMPLATE/new-feature.md) to do so

**Comment by DebugOk** [comment]
I haven't tried compiling this yet, but if you want to clone it to make sure everything works the repo is here https://github.com/DebugOk/Essentials/tree/2.x-ExtendedBaltopAPI

**Pull request merged**